src/main/java/com/minecolonies/coremod/entity/ai/util/AITarget.java

```diff
@@ -24,78 +24,182 @@
+
+    /**
+     * The tickrate at which the Target should be called, e.g. tickRate = 20 means call function every 20 Ticks
+     */
+    @NotNull
+    private        int tickRate;
+    /**
+     * The random offset for Ticks, so that AITargets get more distributed activations on server ticks
+     */
+    @NotNull
+    private final  int tickOffset;
+    /**
+     * The variant used upon creation of the AITarget to uniformly distribute the Tick offset
+     * Static variable counter that changes with each AITarget creation and affects the next one.
+     */
+    private static int tickOffsetVariant = 0;
+
     /**
      * Variable describing if it is okay to eat in a state.
      */
     private boolean okayToEat;
 
     /**
      * Construct a target.
-     *
+     * TODO: Remove once all Targets transitioned to tickRate
      * @param action the action to apply
      */
     public AITarget(@NotNull final Supplier<AIState> action, final boolean isOkayToEat)
     {
-        this(() -> true, isOkayToEat, action);
+        this(() -> true, isOkayToEat, action, 1);
     }
 
     /**
      * Construct a target.
      *
+     * @param action the action to apply
+     */
+    public AITarget(@NotNull final Supplier<AIState> action, final boolean isOkayToEat, @NotNull final int tickRate)
+    {
+        this(() -> true, isOkayToEat, action, tickRate);
+    }
+
+    /**
+     * Construct a target.
+     * TODO: Remove once all Targets transitioned to tickRate
      * @param predicate the predicate for execution
      * @param action    the action to apply
      */
     public AITarget(@NotNull final BooleanSupplier predicate, final boolean isOkayToEat, @NotNull final Supplier<AIState> action)
     {
-        this(null, isOkayToEat, predicate, action);
+        this(null, isOkayToEat, predicate, action, 1);
     }
 
     /**
      * Construct a target.
      *
+     * @param predicate the predicate for execution
+     * @param action    the action to apply
+     */
+    public AITarget(@NotNull final BooleanSupplier predicate, final boolean isOkayToEat, @NotNull final Supplier<AIState> action, @NotNull final int tickRate)
+    {
+        this(null, isOkayToEat, predicate, action, tickRate);
+    }
+
+    /**
+     * Construct a target.
+     * TODO: Remove once all Targets transitioned to tickRate
      * @param state     the state it needs to be | null
      * @param predicate the predicate for execution
      * @param action    the action to apply
      */
-    public AITarget(@Nullable final AIState state, final boolean isOkayToEat, @NotNull final BooleanSupplier predicate, @NotNull final Supplier<AIState> action)
+    public AITarget(
+      @Nullable final AIState state,
+      final boolean isOkayToEat,
+      @NotNull final BooleanSupplier predicate,
+      @NotNull final Supplier<AIState> action,
+      @NotNull final int tickRate)
     {
-        this.state = state;
+        if (state == null)
```

Comment:
why do we allow null?

Reply:
we've been using null for all current AITargets when they weren't limited to a certain state, aka global checks or events. Till stuff is ported over theres an default state for null.